test/controllers/actions/AllowAccessActionSpec.scala

```diff
@@ -304,5 +307,37 @@ class AllowAccessActionSpec
 
       assertEqual(futureResult, None)
     }
+
+    "allow access PSA/PSP permutations" in {
+      val psc: PensionsSchemeConnector = mock[PensionsSchemeConnector]
+
+      def hasAccess(allowPsa: Boolean, allowPsp: Boolean, withSrn: Boolean)(isPsa: Boolean) = {
+        reset(psc)
+        when(psc.checkForAssociation(ArgumentMatchers.eq("A0000000"), any(), ArgumentMatchers.eq(true))(any(), any(), any()))
+          .thenReturn(Future.successful(Right(isPsa)))
+        when(psc.checkForAssociation(ArgumentMatchers.eq("20000000"), any(), ArgumentMatchers.eq(false))(any(), any(), any()))
+          .thenReturn(Future.successful(Right(!isPsa)))
+
+        val psaId = if(isPsa) Some(PsaId("A0000000")) else None
+        val pspId = if(!isPsa) Some(PspId("20000000")) else None
+
+        new TestAllowAccessActionTaskList(if(withSrn) srn else None, psc, allowPsa = allowPsa, allowPsp = allowPsp)
+          .test(OptionalDataRequest(fakeRequest, "id", Some(UserAnswers(Json.obj())), psaId, pspId))
+          .map { case Some(value) => false
+          case None => true
+          }
+      }
+      def test(withSrn: Boolean) = {
+        hasAccess(allowPsa = true, allowPsp = false, withSrn)(isPsa = true).map { hasAccess => assert(hasAccess, true)}
+        hasAccess(allowPsa = true, allowPsp = true, withSrn)(isPsa = true).map { hasAccess => assert(hasAccess, true)}
+        hasAccess(allowPsa = false, allowPsp = true, withSrn)(isPsa = true).map { hasAccess => assert(hasAccess, false)}
+        hasAccess(allowPsa = true, allowPsp = false, withSrn)(isPsa = false).map { hasAccess => assert(hasAccess, false)}
```

Comment:
This isnt exhaustive of the actual possibilities, in order to support this implementation and how it relies and specific combinations in the boolean match can we add the other scenarios. To give a guard until the refactor happens

Reply:
Not sure what is meant by this.
Should I fully simulate a controller with all of the action methods?

Reply:
Introduced false false tests.